File: Leetcode/Python/2962. Count Subarrays Where Max Element Appears at Least K Times/solution.py
# https://leetcode.com/problems/count-subarrays-where-max-element-appears-at-least-k-times/description

class Solution:
    def countSubarrays(self, nums: List[int], k: int) -> int:
        max_num = max(nums)

        count = 0
        max_indexs = []

        for idx, val in enumerate(nums):
            if val == max_num:
                max_indexs.append(idx)
            
            # For each index, add the number of subarrays ending at that index which have exactly k occurrences of the maximum element.
            if len(max_indexs) >= k:
                count += max_indexs[-k] + 1

        return count

File: Leetcode/Typescript/2962. Count Subarrays Where Max Element Appears at Least K Times/solution.ts
// https://leetcode.com/problems/count-subarrays-where-max-element-appears-at-least-k-times/description

function countSubarrays(nums: number[], k: number): number {
  const maxNum = Math.max(...nums);
  let count = 0;
  let queue: number[] = [];

  for (let i = 0; i < nums.length; i++) {
    if (nums[i] === maxNum) {
      queue.push(i);
    }

    // For each index, add the number of subarrays ending at that index which have exactly k occurrences of the maximum element.
    if (queue.length >= k) {
      const startIdx = queue[queue.length - k];
      count += startIdx + 1;
    }
  }

  return count;
}
